### PR TITLE
primitives/x25519: Add x25519 private and public key types

### PR DESCRIPTION
**Pull request by peternose**
It would be nice to have type-safe x25519 private and public key types defined in a similar way as they are in 25519-dalek. Introduced changes follow that direction while using thoroughly tested functions for scalar multiplication.

We could go even one step further to speed up private key functions, but then we would need to refactor/add new tests. Maybe adding types to points and scalars should be done first.
```
type PrivateKey scalar.Scalar
type PublicKey curve.MontgomeryPoint
type SharedSecret curve.MontgomeryPoint

func (priv *PrivateKey) Public() PublicKey {
	var (
		edP   curve.EdwardsPoint
		montP curve.MontgomeryPoint
	)
	montP.SetEdwards(edP.MulBasepoint(curve.ED25519_BASEPOINT_TABLE, (*scalar.Scalar)(priv)))

	return PublicKey(montP)
}

func (priv *PrivateKey) DiffieHellman(pub *PublicKey) SharedSecret {
	var montP curve.MontgomeryPoint
	montP.Mul((*curve.MontgomeryPoint)(pub), (*scalar.Scalar)(priv))

	return SharedSecret(montP)
}
```

### CODE REVIEW

File: primitives/x25519/x25519.go [hunk 1]
```diff
@@ -44,12 +46,84 @@ import (
...
+		rand = cryptorand.Reader
+	}
+
+	var seed [seedSize]byte
+	if _, err := io.ReadFull(rand, seed[:]); err != nil {
+		return nil, err
+	}
+
+	digest := sha512.Sum512_256(seed[:]) // Mitigate poor quality entropy.
+	bzero(seed[:])
```
Comment (Yawning) [inline primitives/x25519/x25519.go:110]: This is largely pointless with Go (and most other languages), omit to be in line with the rest of the package please (Look at what `Sum512_256` does to see that entropy being left in places is inevitable).

**Comment by Yawning** [comment]
I don't really have a strong objection to this per say, as long as the existing API remains so that this is an easy alternative to `x/crypto/curve25519`.

**Review by Yawning — APPROVED** [review]

**Comment by peternose** [comment]
@Yawning Can this be merged? I'm not authorized.

File: primitives/x25519/x25519.go [hunk 2]
```diff
@@ -44,12 +46,83 @@ import (
...
+	return &sec
+}
+
+// PublicKey is the type of X25519 public keys.
+type PublicKey [PublicKeySize]byte
+
+// SharedSecret is the type of the result of a Diffie-Hellman key exchange.
+type SharedSecret [SharedSecretSize]byte
+
+func (ss *SharedSecret) IsZero() bool {
```
Comment (kostko) [inline primitives/x25519/x25519.go:92]: Documentation comment?

**Review by kostko — APPROVED** [review]

**Pull request merged**